proposals/NNNN-multiple-trailing-closures.md

```diff
@@ -0,0 +1,176 @@
+syntax with trailing closure (with disadvantaged described above) or only argument syntax which becomes cumbersome
+and noisy if closures are long or there are too many arguments e.g.
+
+```swift
+when(2 < 3, then: {
+  ...
+  ...
+  ...
+}) {
+  ...
+  ...
+  ...
+}
+```
+
+```swift
+Button(action: {
+  ...
+  ...
+}) {
+  Label("Hello!")
+}
+```
+
+Such syntax for calls is inconsistent, it introduces unnecessary commas and labels.
+
+## Proposed solution
+
+The problem could be fixed by introduction of a uniform spelling for all labeled function arguments as an extension of existing trailing closure feature/syntax e.g.
+
+```swift
+when(2 < 3) {
+   then: {
+     ...
+     ...
+   }
+   else: {
+     ...
+     ...
+   }
+}
+```
+
+or (no parentheses necessary since all arguments are closures)
+
+```swift
+Button {
+  action: {
+    ...
+    ...
+    ...
+  }
+
+  label: {
+    Label("Hello!")
+  }
+}
+```
+
+Proposed new syntax places all labeled closures involved in the call into a single trailing closure block
+that makes it much more human readable and removes a need to delimit calls with commas and parentheses, which is especially important when closures become long.
+
+## Detailed design
+
+Changes required to implement new syntax are constrained solely to the parser, namely to the parsing of call arguments.
+
+It's possible to do an early syntax transformation which would consider each block to be a regular argument and adjust locations of parentheses if necessary.
+For type checker perspective this means that calls with new syntax would just be regular calls which require no special handling:
+
+```swift
+when(2 < 3) {
+   then: { ... }
+   else: { ... }
+}
+```
+
+Would be transformed into:
+
+```swift
+when(2 < 3,
+    then: { ... },
+    else: { ... }
+)
+```
+
+Since all of the essential source information is preserved (locations of all labels and closure blocks) it would be possible for type-checker to produce diagnostics for invalid code without any changes.
+
+## Source compatibility
+
+This is an additive proposal, which makes ill-formed syntax well-formed but otherwise does not affect existing code.
+
+## Effect on ABI stability
+
+This feature is implementable entirely in the parser, as a syntactic transformation on call expressions. It, therefore, has no impact on the ABI.
+
+## Effect on API resilience
+
+Proposed changes do not introduce features that would become a part of a public API.
+
+## Alternatives considered
+
+It has been mentioned on the forums that it's already possible to imitate proposed syntax by placing labeled arguments on separate lines e.g.
+
+```swift
+while(2 < 3,
+  then: {
+    ...
+    ...
+  },
+  else: {
+    ...
+    ...
+  }
+)
+```
+
+This is a reasonable critic but spelling like that feels like a workaround to a requirement for arguments to be comma separated especially when some of them are un-labeled just like `condition` in previous example.
```

Comment:
`critic` ➡️ `critique` (or `criticism`)?

```suggestion
This is a reasonable critique but spelling like that feels like a workaround to a requirement for arguments to be comma separated especially when some of them are un-labeled just like `condition` in previous example.
```

Reply:
Fixed, thank you!